### PR TITLE
Code Request --> Feature/rspec init

### PR DESCRIPTION
**Pull request by rammazzoti2000**
This branch is for code review only.
The Private app, has been built with the user in mind.  Mainly, given the main character (the user), the events and invitations can be created with the user in mind; the user can attend an event, can create and event, can delete, etc.
The main goal was to put in practice the main concepts of Associations in rails:
- create table for user, events, invitation
- user model and events model are directly associated, while invitations is a joined model.
- validated presence of user and events.
- created tests for models with rspec, and integration tests for authorization with capybara and rspec.
- for events managements it has been created integration tests with rspec and capybara

Kindly waiting for your feedback, thank you :wave:

### CODE REVIEW

File: spec/features/event_spec.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+RSpec.describe 'Managing events:', type: :feature do
+  scenario 'create an event --> valid inputs' do
+    visit new_user_url
+    fill_in 'Name', with: 'testing'
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_on 'Submit'
+    visit login_url
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_button 'Beat it'
+    visit new_event_url
```
Comment (shshamim63) [inline spec/features/event_spec.rb:13]: [Optional]
This block is repetitive in your code. You can declare it inside before each to use it in all the scenarios. Something like
```ruby
before(:each) do
    visit new_user_url
    fill_in 'Name', with: 'testing'
    fill_in 'Email', with: 'biscuits@gmail.com'
    click_on 'Submit'
    visit login_url
    fill_in 'Email', with: 'biscuits@gmail.com'
    click_button 'Beat it'
    visit new_event_url
end
```
The following code should appear after line number 4 and you can remove the repetitive chunk from all of your scenarios.

File: spec/features/event_spec.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  scenario 'create an event --> invalid inputs' do
+    visit new_user_url
+    fill_in 'Name', with: 'testing'
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_on 'Submit'
+    visit login_url
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_button 'Beat it'
+    visit new_event_url
```
Comment (shshamim63) [inline spec/features/event_spec.rb]: Repeated code.

File: spec/features/event_spec.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    expect(page).to have_content("Title can't be blank")
+    expect(page).to have_content("Description can't be blank")
+  end
+
+  scenario 'edit an event --> valid inputs' do
+    visit new_user_url
+    fill_in 'Name', with: 'testing'
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_on 'Submit'
+    visit login_url
```
Comment (shshamim63) [inline spec/features/event_spec.rb]: Repeated code.

File: spec/features/event_spec.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  scenario 'delete an event' do
+    visit new_user_url
+    fill_in 'Name', with: 'testing'
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_on 'Submit'
+    visit login_url
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_button 'Beat it'
+    visit new_event_path
```
Comment (shshamim63) [inline spec/features/event_spec.rb]: Repeated code.

File: spec/features/event_spec.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+  scenario 'show an event' do
+    visit new_user_url
+    fill_in 'Name', with: 'testing'
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_on 'Submit'
+    visit login_url
+    fill_in 'Email', with: 'biscuits@gmail.com'
+    click_button 'Beat it'
+    visit new_event_url
```
Comment (shshamim63) [inline spec/features/event_spec.rb]: Same repeated code.

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: : Required Changes :1234: : 
___
Hi @rammazzoti2000 and @vzdrizhni :wave: 
Awesome job :+1: Almost near to complete this project. I just found a crucial task that you left behind. Let's cover it.
___
### :spades: Mandatory Changes :spades: 
1. The user show method should be related to the current user. But your one is rendering data based on the user_id params. It should not. This method should render the following info
  - lists events created by current user ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  - lists events previoulsy attendented by current user ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  - lists upcoming events previously attended by current user ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
___
### :clubs: Optional Changes :clubs: 
___
- You can add `before each` block to run a sudden amount of the same code before each of the tests. I have added a sample code.
___
##### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

File: spec/features/event_spec.rb (hunk 1 above)
Comment (rammazzoti2000) [reply spec/features/event_spec.rb:13]: thank you

**Comment by rammazzoti2000** [comment]
Fixed changes required by the TSE:
- [x]  The user show method should be related to the current user, in particular lists events created by current user
- [x]  Add `before each` block to run a sudden amount of the same code before each of the tests

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,13 +1,53 @@
...
+  def edit; end
+
+  # POST /users
+  # POST /users.json
   def create
+    @user = User.new(user_params)
 
+    respond_to do |format|
+      if @user.save
+        format.html { redirect_to root_path, notice: "User was successfully created. #{@user.name} please login." }
```
Comment (mosaaleb) [inline app/controllers/users_controller.rb]: - [x] **[OPTIONAL]** For better user experience, let's sign in the user automatically after creation.

File: app/views/events/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    </thead>
+
+    <tbody>
+      <% @upcoming_events.each do |event| %>
+        <tr class="box box-radius">
+          <td class="column">Event: <%= link_to event.title, event %></td>
+          <td class="column"><div class="description">Description: <%= event.description %><div></td>
+          <td class="column">Date: <%= event.date.to_formatted_s(:long_ordinal)  %></td>
+          <td class="column">Creator: <%= event.creator.name %></td>
+          <td class="column"><%= link_to 'Delete', event, method: :delete, class:'button is-primary has-text-white'  %></td>
```
Comment (mosaaleb) [inline app/views/events/index.html.erb]: - [x] Any user can delete any event even if he/she is not the creator of the event, let's add a check to render the delete button only if the current user is the same as the event creator. (Don't forget to wrap this logic in events helper module `app/helpers/`).

File: spec/features/event_spec.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+  scenario 'show an event' do
+    fill_in 'Title', with: 'Minneapolis'
+    fill_in 'Description', with: 'Mama are mere, Tata are pere'
+    click_button 'Create'
+    visit root_path
+    click_on('Minneapolis')
+    expect(page).to have_content('Minneapolis')
+    expect(page).to have_content('Mama are mere, Tata are pere')
+  end
```
Comment (mosaaleb) [inline spec/features/event_spec.rb:61]: - [x] Let's add a test for attending an event. (when user is not attending and when user is already attending)

I also suggest to take a look at [better specs](http://www.betterspecs.org/) on how to organize your tests.

File: spec/models/invitation_spec.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+RSpec.describe Invitation, type: :model do
+  describe 'associations' do
+    it { should belong_to(:attendee).with_foreign_key(:attendee_id).class_name('User') }
+    it { should belong_to(:attended_event).with_foreign_key(:attended_event_id).class_name('Event') }
+  end
+
+  # rubocop:disable Layout/LineLength
+  describe 'validations' do
+    it { should validate_uniqueness_of(:attended_event_id).case_insensitive.scoped_to(:attendee_id).with_message('You can only atend once (Matrix! Perhaps!?)') }
```
Comment (mosaaleb) [inline spec/models/invitation_spec.rb:11]: :clap:

**Review by mosaaleb — CHANGES_REQUESTED** [review]
#### Excellent job @rammazzoti2000 @vzdrizhni :clap:

Nice job handling the requirements requested by my fellow reviewer, still there is one major bug (any user can delete any event)

- [x] **[OPTIONAL]** We can see that we have lots of duplicated code in `users/show.html.erb` file, and since the structure of a "single event" whether it's past or previous or created by the user is the same, we can create a partial file for event `events/_event.html.erb` to dry out our code a little.
[using partials](https://guides.rubyonrails.org/layouts_and_rendering.html#using-partials) 
[collection rendering](https://guides.rubyonrails.org/layouts_and_rendering.html#rendering-collections)

Our code after refactoring can like like:

```erb
<%= render @events_by_current_user %>
<%= render @upcoming_events %>
<%= render @prev_events %>
```

Please handle this requirement and submit a new code review.

##### Muhammad Ebeid

**Comment by rammazzoti2000** [comment]
Fixed all changes required by TSE:
- Use of partials for events view
- Tests for attending or not an event
- Helper method to delete only events of current user
- User is signed in automatically after creation

**Review by mosesogwo — APPROVED** [review]
**Great! Project is approved!**

**Cheers:stars:**
### Status: Approved:heavy_check_mark:

**Pull request merged**